### PR TITLE
[Build] Code refactoring for AutoPatching

### PR DESCRIPTION
**Pull request by shantanutrip**
*GitHub Issue #, if available:*

**Note**: 
- If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

- All PR's are checked weekly for staleness. This PR will be closed if not updated in 30 days.

### Description
This PR refactors AutoPatch code to add the logic for layer optimization.

### Tests run

**NOTE: By default, docker builds are disabled. In order to build your container, please update dlc_developer_config.toml and specify the framework to build in "build_frameworks"**
- [ ] I have run builds/tests on commit <INSERT COMMIT ID> for my changes.

**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

### Formatting
- [ ] I have run `black -l 100` on my code (formatting tool: https://black.readthedocs.io/en/stable/getting_started.html)

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: miscellaneous_scripts/dockerfile_patch_script.sh [hunk 1]
```diff
@@ -2,32 +2,44 @@
 set -e
 
 RELEASED_IMAGE_SHA=$1
+patching_info_path=/opt/aws/dlc/patching-info
```
Comment (shantanutrip) [inline miscellaneous_scripts/dockerfile_patch_script.sh:5]: Change this variable name to CAPS

File: miscellaneous_scripts/dockerfile_patch_script.sh [hunk 2]
```diff
@@ -2,32 +2,44 @@
 set -e
 
 RELEASED_IMAGE_SHA=$1
```
Comment (shantanutrip) [inline miscellaneous_scripts/dockerfile_patch_script.sh:4]: LATEST_RELEASED_IMAGE_SHA

File: src/patch_helper.py [hunk 3]
```diff
@@ -148,37 +154,81 @@ def conduct_autopatch_build_setup(pre_push_image_object: DockerImage, download_p
...
+        get_cloned_folder_path(),
+        f"""{base_image_uri_for_patch_builds.replace("/", "_").replace(":", "_")}_patch-dump""",
     )
-    if not os.path.exists(patch_details_path):
-        run(f"mkdir {patch_details_path}", hide=True)
+    if not os.path.exists(complete_patching_info_dump_location):
+        run(f"mkdir {complete_patching_info_dump_location}", hide=True)
 
-    run(f"docker pull {released_image_uri}", hide=True)
+    FORMATTER.print(f"[current_patch_details_path] {current_patch_details_path}")
```
Comment (shantanutrip) [inline src/patch_helper.py:199]: Remove

File: src/patch_helper.py [hunk 4]
```diff
@@ -265,7 +316,9 @@ def retrive_autopatched_image_history_and_upload_to_s3(image_uri):
     container_id = run(f"{docker_run_cmd}", hide=True).stdout.strip()
     try:
         docker_exec_cmd = f"docker exec -i {container_id}"
-        history_retrieval_command = f"cat /opt/aws/dlc/patch-details/overall_history.txt"
+        history_retrieval_command = (
+            f"cat /opt/aws/dlc/patching-info/patch-details/overall_history.txt"
```
Comment (shantanutrip) [inline src/patch_helper.py:320]: Use global variables

File: src/safety_report_generator.py [hunk 5]
```diff
@@ -131,9 +131,7 @@ def get_autopatched_dumped_ignore_dict_of_packages(self):
         """
         This method extracts the dumped ignore lists within the DLCs that have been dumped by the autopatch procedure.
         """
-        dumped_ignore_list_command = (
-            f"{self.docker_exec_cmd} cat /opt/aws/dlc/patch-details/vuln_deactivation_data.json"
-        )
+        dumped_ignore_list_command = f"{self.docker_exec_cmd} cat /opt/aws/dlc/patching-info/patch-details/vuln_deactivation_data.json"
```
Comment (shantanutrip) [inline src/safety_report_generator.py:134]: Make a constant

File: src/utils.py [hunk 6]
```diff
@@ -512,3 +512,87 @@ def get_dummy_boto_client():
     :return: BotocoreClientSTS
     """
     return boto3.client("sts", region_name=os.getenv("REGION"))
+
+
+def get_folder_size_in_bytes(folder_path):
```
Comment (shantanutrip) [inline src/utils.py:517]: Add a comment

File: test/test_utils/__init__.py [hunk 7]
```diff
@@ -2235,3 +2244,54 @@ def get_installed_python_packages_with_version(docker_exec_command: str):
...
+    corresponding_image_spec = None
+
+    for _, image_spec in buildspec_def["images"].items():
+        if image_tag.startswith(image_spec["tag"]):
+            if corresponding_image_spec:
+                raise ValueError(
+                    f"""Mutliple tags - {corresponding_image_spec["tag"]}, {image_spec["tag"]} found for the image {image_uri}"""
+                )
+            corresponding_image_spec = image_spec
+
```
Comment (shantanutrip) [inline test/test_utils/__init__.py:2296]: Raise an error here in case nothing matches

File: test/dlc_tests/sanity/test_pre_release.py [hunk 8]
```diff
@@ -1028,3 +1035,69 @@ def test_core_package_version(image):
...
+        if package_name not in current_image_package_version_dict:
+            violating_packages[
+                package_name
+            ] = "Not present in the image that is being currently built."
+            continue
+        version_in_current_image = current_image_package_version_dict[package_name]
+        if Version(version_in_released_image) > Version(version_in_current_image):
+            violating_packages[
+                package_name
+            ] = f"Version in already relased image: {version_in_released_image} is greater that version in current image: {version_in_current_image}"
```
Comment (shantanutrip) [inline test/dlc_tests/sanity/test_pre_release.py:1099]: nit: `released`

File: src/utils.py (hunk 6 above)
Comment (saimidu) [reply src/utils.py:517]: about why `du -sh` doesn't work here.

File: test/dlc_tests/sanity/test_ecr_scan.py [hunk 9]
```diff
@@ -143,6 +147,8 @@ def helper_function_for_leftover_vulnerabilities_from_enhanced_scanning(
 
     :param image: str Image URI for image to be tested
     :param python_version: str, This parameter is used for extracting allowlist for canary image uris that do not have a python version in it.
+    :param minimum_sev_threshold: str, If minimum_sev_threshold is set vulnerabilities with severity < minimum_sev_threshold will not be taken into consideration.
+    :param allowlist_removal_enabled: boolean, Value of this parameter decides if we should remove allowlisted vulnearbilities from the scanner results.
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_ecr_scan.py:151]: Please add docstring for `remove_non_patchable_vulns`

File: test/dlc_tests/sanity/test_ecr_scan.py [hunk 10]
```diff
@@ -143,6 +147,8 @@ def helper_function_for_leftover_vulnerabilities_from_enhanced_scanning(
...
     :return: ECREnhancedScanVulnerabilityList Object with leftover vulnerability data
     :return: ecr_enhanced_repo_uri, String for the image uri in the enhanced scanning repo
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_ecr_scan.py:153]: Is the return value in the docstring in line 152 incorrect?

File: test/dlc_tests/sanity/test_pre_release.py [hunk 11]
```diff
@@ -1028,3 +1035,69 @@ def test_core_package_version(image):
...
         not violation_data
     ), f"Few packages violate the core_package specifications: {violation_data}"
+
+
+@pytest.mark.model("N/A")
+def test_package_version_regression_in_image(image):
+    """
+    This test verifies if the python package versions in the already released image are not being downgraded/deleted in the
+    new released. This test would be skipped for images whose BuildSpec does not have `latest_release_tag` or `release_repository`
+    keys in the buildspec - as these keys are used to extract the released image uri. Additionally, if the image is not already
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_pre_release.py:1045]: `test_package_version_regression_in_image` seems to be passing on PT 2.1 inference even though the buildspec doesn't have the `latest_release_tag` or `release_repository` keys.

**Review by saimidu — APPROVED** [review]

File: src/utils.py (hunk 6 above)
Comment (shantanutrip) [reply src/utils.py:533]: Added the comment. Thanks!

File: test/dlc_tests/sanity/test_ecr_scan.py (hunk 9 above)
Comment (shantanutrip) [reply test/dlc_tests/sanity/test_ecr_scan.py]: Added the docstring. Thanks!!

**Review by saimidu — APPROVED** [review]

**Review by saimidu — APPROVED** [review]

**Comment by saimidu** [comment]
Could you please make the PR title and the commit message on master more specific as to what exactly is being refactored, or why the refactor is being done?

**Pull request merged**